File: days/4/puzzles/1/program.js
const MIN = 136760;
const MAX = 595730;

let numValidPasswords = 0;

for (let i = MIN; i <= MAX; i++) {
  const password = i;

  if (isValidPassword(password)) {
    numValidPasswords += 1;
  }
}

console.log(numValidPasswords);

function isValidPassword(password) {
  let hasDouble = false;
  let neverDecreasing = true;
  const digits = getDigits(password);

  for (let i = 0; i < digits.length; i++) {
    if (typeof digits[i+1] !== 'undefined' && digits[i+1] === digits[i]) {
      hasDouble = true;
    }

    if (typeof digits[i+1] !== 'undefined' && digits[i+1] < digits[i]) {
      neverDecreasing = false;
    }
  }

  return hasDouble && neverDecreasing;
}

function getDigits(number) {
  return (number+'').split('').map(digit => parseInt(digit, 10));
}

File: days/4/puzzles/1/program.py
MIN = 136760
MAX = 595730

def is_valid_password(password):
  has_double = False
  never_decreasing = True
  digits = [int(d) for d in str(password)]

  for n, nth_digit in enumerate(digits):
    if n + 1 < len(digits):
      if digits[n+1] == nth_digit:
        has_double = True

      if digits[n+1] < nth_digit:
        never_decreasing = False

  return (has_double and never_decreasing)

num_valid_passwords = 0

for i in range(MIN, MAX + 1):
  if is_valid_password(i):
    num_valid_passwords += 1

print(num_valid_passwords)
